In [20]:
import requests
from bs4 import BeautifulSoup
import re

In [12]:

# ucla
url = "https://www.ratemyprofessors.com/school?sid=1381"

response = requests.get(url)
html = response.text

soup = BeautifulSoup(html, "html.parser")

school_name = soup.find("div", {"class": "HeaderDescription__StyledTitleName-sc-1lt205f-1 eNxccF"}).text.strip()
overall_score = soup.find("div", {"class": "OverallRating__Number-y66epv-3 dXoyqn"}).text.strip()
reputation = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Reputation").find_next_sibling().text.strip()
location = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Location").find_next_sibling().text.strip()
facilities = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Facilities").find_next_sibling().text.strip()
food = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Food").find_next_sibling().text.strip()
happiness = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Happiness").find_next_sibling().text.strip()
opportunities = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Opportunities").find_next_sibling().text.strip()
clubs = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Clubs").find_next_sibling().text.strip()
safety = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Safety").find_next_sibling().text.strip()
social = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Social").find_next_sibling().text.strip()
internet = soup.find("div", {"class": "CategoryGrade__CategoryTitle-sc-17vzv7e-1 XKroK"}, text="Internet").find_next_sibling().text.strip()


# Print the overall quality score
print("School Name:", school_name)
print("Overall quality score:", overall_score)
print("Reputation score", reputation)
print("Location score", location)
print("Facilities score", facilities)
print("Food score", food)
print("Happiness score", happiness)
print("Opportunities score", opportunities)
print("Clubs score", clubs)
print("Safety clubs", safety)
print("Social score", social)
print("Internet score", internet)

School Name: University of Southern California
Overall quality score: 4.4
Reputation score 4.6
Location score 3.9
Facilities score 4.6
Food score 4.0
Happiness score 4.6
Opportunities score 4.7
Clubs score 4.5
Safety clubs 3.5
Social score 4.6
Internet score 4.1


In [13]:
def extract_comment(text):
    pattern = r'SchoolRating__RatingComment[^>]*>([^<]*)'
    match = re.search(pattern, text)
    if match:
        return match.group(1)
    else:
        return None


In [14]:
def extract_awesomeScore(text):
    pattern = r'awesomeScore[^>]*>([^<]*)'
    match = re.search(pattern, text)
    if match:
        return match.group(1)
    else:
        return None

In [15]:
def extract_review_date(text):
    pattern = r'TimeStamp__StyledTimeStamp-sc-9q2r30-0 bXQmMr SchoolRating__StyledTimeStamp-sb9dsm-7 bkDMlg[^>]*>([^<]*)'
    match = re.search(pattern, text)
    if match:
        return match.group(1)
    else:
        return None

In [16]:
def count_approved_substrings(input_string):
    approved_strings = ["DisplaySlider__DisplaySliderBox-sc-6etfq5-3 gjzHHH", "DisplaySlider__DisplaySliderBox-sc-6etfq5-3 jcnvup",
                        "DisplaySlider__DisplaySliderBox-sc-6etfq5-3 kPgDgT", "DisplaySlider__DisplaySliderBox-sc-6etfq5-3 bXxFBs",
                        "DisplaySlider__DisplaySliderBox-sc-6etfq5-3 ggogdh", "DisplaySlider__DisplaySliderBox-sc-6etfq5-3 ffSfbF",
                        "DisplaySlider__DisplaySliderBox-sc-6etfq5-3 idrjOK"]
    count = 0
    for approved_string in approved_strings:
        count += input_string.count(approved_string)
    return count


In [17]:
def extract_review_metric_and_score(text):
    scores = []
    slider_divs = text.find_all("div", {"class": "DisplaySlider__SliderBoxContainer-sc-6etfq5-2 LReVN"})

    for slider_div in zip(slider_divs):
        scores.append(count_approved_substrings(str(slider_div)))
    
    return scores
    


In [18]:
# don't do this repeatedly
if False:
    response = requests.get(url)
    content = response.text
    soup = BeautifulSoup(content, 'html.parser')
    ratings = soup.find_all('div', class_='SchoolRating__SchoolRatingContainer-sb9dsm-0 inMLDw')


In [21]:
for rating in ratings:

    comment = extract_comment(str(rating))
    print("Comment:", comment)

    awesomeScore = extract_awesomeScore(str(rating))
    print("Awesome Score for this user:", awesomeScore)

    review_date = extract_review_date(str(rating))
    print("Review Date:", review_date)

    metric = extract_review_metric_and_score(rating)
    print()


Comment: Reputation- Excellent and still improving, impresses people I’ve talked to all over the US. Opportunities- the network can get you into any industry you want. Facilities- I love the campus, so nice to walk around and so many cool study spots. Social- there’s never a dull moment unless you want one.
Awesome Score for this user: 4.6
Review Date: Feb 24th, 2023
Reputation: 5
Location: 3
Opportunities: 5
Facilities: 5
Internet: 5
Food: 4
Clubs: 5
Social: 5
Happiness: 5
Safety: 4

Comment: USC is the number one university and their School of Dramatic Arts is revered around the world. I don't even go here, and even I know that USC is #1. Trojans Rule and always will.
Awesome Score for this user: 4.9
Review Date: Feb 6th, 2023
Reputation: 5
Location: 4
Opportunities: 5
Facilities: 5
Internet: 5
Food: 5
Clubs: 5
Social: 5
Happiness: 5
Safety: 5

Comment: I love this school
Awesome Score for this user: 4.6
Review Date: Feb 3rd, 2023
Reputation: 5
Location: 4
Opportunities: 5
Facilities

|              | Green                                              | Yellow                                             | Red                                                | Gray                                               |
|--------------|----------------------------------------------------|----------------------------------------------------|----------------------------------------------------|----------------------------------------------------|
| First        | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 gjzHHH | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 jcnvup | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 kPgDgT |                                                    |
| Middle (2-4) | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 bXxFBs | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 ggogdh | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 ffSfbF | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 jmBUzm |
| End          | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 idrjOK |                                                    |                                                    | DisplaySlider__DisplaySliderBox-sc-6etfq5-3 kjbVpX |

{
    "USC" : {
        "School_Overall" : 4.5,
        "School_Reputation" : 4.0,
        "School_Location" : 4.5,
        "School_Facilities" : 3.7,
        "School_Food" : 4.3,

        "Comments" : {
            "This place is nice" : {
                "Comment_Date": "Feb 23rd, 2023"
                "Comment_Overall" : 4,
                "Comment_Reputation" : 3,
                "Comment_Location" : 2,
                "Comment_Facilities" : 4,
                "Comment_Food" : 5
            }
        }
    }
}